### PR TITLE
choleskey code serial/mdspan/stadpar

### PR DESCRIPTION
**Pull request by hcq9102**
- [x] step1: serial code works& verified correctness 
            
        - And add mdspan in serial code

- [x] step2: add stdpar & verified correctness
     
       Note:  for -DSTDPAR=gpu case, careful with lambda function by reference.

Add command options.

### CODE REVIEW

File: apps/choleskey/choleskey_stdpar.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ 
+    // Decomposing a matrix into Lower Triangular
+    for (int i = 0; i < matrix.extent(0); i++) {
+        for (int j = 0; j <= i; j++) {
+            int sum = 0;
+ 
+            if (j == i) // summation for diagonals
+            {
+                std::for_each_n(std::execution::par, counting_iterator(0), j,
+                      [=, j = j, &sum](int32_t k) {
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:27]: for `-DSTDPAR=gpu` case, I wonder can we successfully pass `sum` variable to lambda function by reference? Will we encounter memory access error since `sum` has not allocated on GPU?

File: apps/choleskey/choleskey_stdpar_snd.cpp [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+{
+    constexpr int n = 3;
+    std::vector<int> v{4, 12, -16,12, 37, -43, -16, -43, 98 };
+    Cholesky_Decomposition(v, n);
+
+    return 0;
+}
+
+
+// auto Cholesky_Decomposition_async(int elem){
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar_snd.cpp:88]: In case you are trying to add async to this PR, let's move the sender implementation in a separate PR.

**Review by weilewei — COMMENTED** [review]
Please take a look on my comments.

File: apps/choleskey/choleskey_stdpar.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ 
+// Driver Code for testing
+int main()
+{
+    constexpr int n = 3;
+    std::vector<int> v{4, 12, -16,12, 37, -43, -16, -43, 98 };
+    Cholesky_Decomposition(v, n);
+
+    return 0;
+}
```
[the remark this answers is not in the document]
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp:69]: Sure! will do

File: apps/choleskey/choleskey_stdpar_snd.cpp (hunk 2 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar_snd.cpp:88]: okay.

File: apps/choleskey/choleskey_stdpar.cpp (hunk 1 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp:27]: I have tried to pass `sum` by value to avoid the memory warning, but got error issue:

> error: expression must be a modifiable lvalu
 sum += pow(lower[j][k], 2);
   ^

Comment (weilewei) [reply apps/choleskey/choleskey_stdpar.cpp:27]: how about using `transform_reduce` if this works for your purpose. 
```
                sum = std::transform_reduce(std::execution::par_unseq, lower[j].cbegin(), lower[j].cend(), 
                    sum, std::plus{},
                        [=, j = j](auto k) { return pow(lower[j][k], 2); });
```

Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp:27]: this perfectly resolved the capture issue, thanks

File: apps/choleskey/choleskey_stdpar_snd.cpp
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar_snd.cpp:1]: please remove this file from this PR. also, edit the PR description for bulk/sender item accordingly.

File: apps/choleskey/choleskey_stdpar.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#include <experimental/mdspan>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+using namespace std;
+
+struct Multiply {
+  int operator()(int a, int b) const { return a * b; }
+};
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:16]: nit: prefer to have it as lambda function. Also, it seems it can be generalized and used by line 39 (the n*n).

File: apps/choleskey/choleskey_stdpar.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  // Decomposing a matrix into Lower Triangular
+  for (int i = 0; i < matrix.extent(0); i++) {
+    for (int j = 0; j <= i; j++) {
+      int sum = 0;
+
+      if (j == i)  // summation for diagonals
+      {
+
+        // apply transform_reduce for the first [lower[j].cbegin(), lower[j].cbegin() +j )
+        // sum = std::transform_reduce(std::execution::par, lower[j].cbegin(), lower[j].cbegin() +j, 0, std::plus{}, Square());
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:35]: nit: remove comments in the finalized PR.

File: apps/choleskey/choleskey_stdpar.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    cout << endl;
+  }
+}
+
+// Driver Code for testing
+int main() {
+  constexpr int n = 3;
+  std::vector<int> v{4, 12, -16, 12, 37, -43, -16, -43, 98};
+  Cholesky_Decomposition(v, n);
+
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:76]: Please make the main can accept generalized shape. Use `argparse::Args`, see other examples how to configure the shape of array dimension. For example: https://github.com/mhaseeb123/nvstdpar/blob/main/apps/1d_stencil/stencil_stdpar.cpp#L36-L51.

Similarly, make the printouts or `cout` statements as optional. Ultimately, we would like to see runtime for large array size. Also refer to other examples.

**Review by weilewei — COMMENTED** [review]
Overall looks great! Please see comments inline. Once addressed, we can merge it. Thanks!

File: apps/choleskey/choleskey_stdpar.cpp (hunk 6 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp]: okay. done.

File: apps/choleskey/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+target_include_directories(
+  choleskey_stdpar
+  PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_LIST_DIR}/../../include
+          ${ARGPARSE_INCLUDE_DIR} ${MDSPAN_INCLUDE_DIR})           
+
+add_executable(choleskey_stdpar_foreach choleskey_stdpar_foreach.cpp)
+target_include_directories(
+  choleskey_stdpar_foreach
+  PRIVATE ${CMAKE_BINARY_DIR} ${CMAKE_CURRENT_LIST_DIR}/../../include
+          ${ARGPARSE_INCLUDE_DIR} ${MDSPAN_INCLUDE_DIR})
```
Comment (weilewei) [inline apps/choleskey/CMakeLists.txt:19]: remove these lines, since this PR does not have `choleskey_stdpar_foreach.cpp`

File: apps/choleskey/choleskey_stdpar.cpp (hunk 4 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp]: sure. done

File: apps/choleskey/choleskey_stdpar_snd.cpp
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar_snd.cpp:1]: will do. thank  you.

File: apps/choleskey/choleskey_stdpar.cpp (hunk 5 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp]: yes!

File: apps/choleskey/choleskey_stdpar.cpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  // see if help wanted
+  if (args.help) {
+    args.print();  // prints all variables
+    return 0;
+  }
+
+  benchmark(args);
+
+  return 0;
+}
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:129]: `\ No newline at end of file` See CI results: https://github.com/mhaseeb123/nvstdpar/actions/runs/6266675776/job/17018364471?pr=26

File: apps/choleskey/choleskey_stdpar_snd.cpp
Comment (weilewei) [reply apps/choleskey/choleskey_stdpar_snd.cpp:1]: remove this file.

File: apps/choleskey/matrixutil.hpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+//     }
+//   }
+
+//   std::vector<T> flattenedVector;
+//   for (const auto& row : matrix) {
+//     flattenedVector.insert(flattenedVector.end(), row.begin(), row.end());
+//   }
+
+//   return flattenedVector;
+// }
```
Comment (weilewei) [inline apps/choleskey/matrixutil.hpp:93]: remove these commented lines.

**Review by weilewei — COMMENTED** [review]
The PR looks good to me, some formatting and clean up are needed.

File: apps/choleskey/choleskey_stdpar_snd.cpp
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar_snd.cpp:1]: done

File: apps/choleskey/choleskey_stdpar.cpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    // Decomposing a matrix into Lower Triangular
+    for (int i = 0; i < matrix_ms.extent(0); i++) {
+      for (int j = 0; j <= i; j++) {
+        T sum = 0;
+
+        if (j == i)  // summation for diagonals
+        {
+          sum = std::transform_reduce(
+              std::execution::par, lower[j].cbegin(), lower[j].cbegin() + j, 0,
+              std::plus{}, [=, i = i, j = j](int val) { return val * val; });
```
Comment (weilewei) [inline apps/choleskey/choleskey_stdpar.cpp:54]: It seems the i and j variables are not used in the lambda function. If true, we can remove it.

**Review by weilewei — APPROVED** [review]
One more nit change

File: apps/choleskey/choleskey_stdpar.cpp (hunk 10 above)
Comment (hcq9102) [reply apps/choleskey/choleskey_stdpar.cpp:54]: okay. Thanks.

**Comment by weilewei** [comment]
LGTM! Thanks!
